0009-comit-btsieve-db.adoc

```diff
@@ -0,0 +1,245 @@
+
+== Goals
+
+1. No separate process
+
+One of the core development goals of COMIT is to be as easily accessible as possible.
+Hence, any solution that involves running a _separate_ process for persistently storing data is not considered.
+All the solutions proposed here need to be embeddable into `comit_node` and `btsieve`.
+
+2. Atomic access
+
+Data integrity is key to our software.
+Hence, all access to the data needs be done in an atomic fashion.
+
+== What do we want to store?
+
+In order to decide on the best library/solution for storing data in `comit_node` and `btsieve` we need to know, what data we want to store and how we would like to store/access it.
+
+=== At the time of writing, we want to store the following things:
+
+==== `comit_node`
+
+- A list of queries created on `btsieve`
+- A list of swaps +
+This sounds more trivial than it is.
+The data structure for a swap is different for each protocol (rfc003 has expiry fields, secret hash etc).
+In addition, each swap might have different ledgers which in turn have different parameters and hence a different shape.
+
+==== `btsieve`
+
+- The list of created queries
+- Retrieved blocks (probably at maximum for the last 24h?)
+- Metadata related to each blockchain +
+For example, the latest blockhash at the time of startup and probably other stuff I can't think of right now.
+
+=== Things that we may want to store in the future:
+
+- Addresses and PeerIDs of peers we have already seen in the past
+- PeerIDs we don't want to communicate with (banned/blocked bc of bad behaviour)
+- Our own external address as observed by other parties
+
+=== Key/Value vs SQL vs document storage
+
+Before diving in the comparison of the different database engines, let's have a look at what we would have to do in order for the three main types of storage APIs:
+
+==== Key/Value
+
+This is the most basic type of interface for a storage backend.
+Key/Value stores usually expose a `schema-on-read` kind of interface, that is, you can store whatever you want under a given key and you need to know the schema of the data upon retrieval.
+To use this for our needs, we would have to create our own namespacing when creating the keys.
+For the swaps for example, we would have to something along the lines of:
+```
+key: "swaps.rfc003.<swapId>"
+value: bytes of the JSON serialization of the swap struct
+```
+This + the metadata of the swap would allow us to know the shape of the value upfront.
+
+To summarize, for all Key/Value based storage solutions, we would have to:
+
+- decide on a namespacing scheme that allows us to separate values of different shape from each other (kind of mimicing SQL tables)
+- decide on a serialization format for the values (Key/Value stores usually only provide a `bytes` -> `bytes` kind of mapping, i.e. the key aswell as the value are arbitrary `bytes`)
```

Comment:
Also decide on versioning can be quite challenging as there are various possible strategies:
- version byte in value (or even in key!)
- new fields are always optional
- try and fail deserialization